content/configs/config_file.md
Значення потрібно редагувати в [configs.ini](/hsc_gov_subscriber/config.ini).

Для "Практичний іспит (**транспортний засіб навчального закладу**)" змінюється тільки SUB_ID в залежності від категорії
* на категорії A1; A; B1 — QUESTION_ID: 56, SUB_ID: 12
* на категорії B; BE — QUESTION_ID: 56, SUB_ID: 13
* на категорії C1; C — QUESTION_ID: 56, SUB_ID: 14
* на категорії D1; D — QUESTION_ID: 56, SUB_ID: 15
* на категорії C1E; CE — QUESTION_ID: 56, SUB_ID: 16
* на категорії D1E; DE — QUESTION_ID: 56, SUB_ID: 17

Для "Практичний іспит (**транспортний засіб сервісного центру**)" - використовується тільки QUESTION_ID (без SUB_ID)
* на категорію А1 — QUESTION_ID: 59
* на категорію А — QUESTION_ID: 60
* на категорію В1 — QUESTION_ID: 61
* на категорію В (з механічною коробкою передач) — QUESTION_ID: 55
* на категорію В (з автоматичною коробкою передач) — QUESTION_ID: 58
* на категорію С1 — QUESTION_ID: 62
* на категорію С — QUESTION_ID: 63
* на категорію D1 — QUESTION_ID: 64
* на категорію D — QUESTION_ID: 65
* на категорію BE — QUESTION_ID: 66
* на категорію C1E — QUESTION_ID: 67
* на категорію CE — QUESTION_ID: 68
* на категорію D1E — QUESTION_ID: 69
* на категорію DE — QUESTION_ID: 70

`email` - пошта на яку прийде білет та нагадування.

`office_id` - ID офісу ГСЦ. (71 - м. Миколаїв, пров. Транспортний, 1а/1)

`question_id` - ID послуги. (56 - практичний іспит (транспортний засіб навчального закладу).

`school_vehicle_practical_exam` - Який транспортний засіб на практичний іспит. (true - Практичний іспит (транспортний засіб **навчального закладу**), false - "Практичний іспит (транспортний засіб **сервісного центру**)")  

`vin` - Номерний знак або VIN(останні 6 цифр). (обозявковий якщо `question_id` = 49)

`api_id` - `App api_id` з [п.1](/content/configs/configuring.md).

`api_hash` - `App api_hash` з [п.1](/content/configs/configuring.md).

`start_date` - дата з якої потрібно брати талон. (наприклад 2024-05-9 - якщо буде талон на 08.5.2024 - код не візьме
такий талон, якщо прийде 09.5.2024 - код намагатиметься взяти талон)

`end_date` - дата з якої вже не потрібно брати талон. (наприклад 2024-05-14 - якщо буде талон на 15.5.2024 - код не
візьме такий талон, якщо прийде 14.5.2024 або 13.5.2024 - код намагатиметься взяти талон)

Гайд [де брати `OFFICE_ID` та `QUESTION_ID`](/content/configs/browser_requests/pract_ispt_id.md).

### Приклад конфіг файлу для `практичний іспит (транспортний засіб навчального закладу) на категорії B; BE;`

![alt text](/content/configs/photo_2024-10-06_17-51-32.jpg)


### Приклад конфіг файлу для `Практичний іспит на категорію В (транспортний засіб сервісного центру з автоматичною коробкою передач)`

![alt text](/content/configs/photo_2024-10-06_18-15-53.jpg)



content/configs/configuring.md
## Налаштування

<details>
<summary>1. Телеграм</summary>

Для прослуховування телеграма потрібні `App api_id` та `App api_hash`.

Гайд [як створити `App api_id` та `App api_hash`](/content/configs/tg/tg_api.md).

</details>

<details>
<summary>2. Конфіг файл</summary>

Опис конфіг файлу [config_file](/content/configs/config_file.md).

Гайд [де брати `OFFICE_ID` та `QUESTION_ID`](/content/configs/browser_requests/pract_ispt_id.md).

</details>

<details>
<summary>3. Файли cookies</summary>

Після вдалої авторизації на [https://eq.hsc.gov.ua/](https://eq.hsc.gov.ua/) потрібно додати кукі
в [cookies.json](/hsc_gov_subscriber/cookies.json) для авторизованих запитів на сайт.

**Кукі злітають автоматом через 30хв якщо не оновлювати сторінку сайту.** Щоб цього уникнути
рекомендую [встановити розширення бля браузеру](https://chromewebstore.google.com/detail/easy-auto-refresh/aabcgdmkeabbnleenpncegpcngjpnjkc?hl=en-US&utm_source=ext_sidebar)
і виставити 300сек.

Опис cookies файлу [cookies.json](/content/configs/cookies.md).

Гайд [де брати `cookies`](/content/configs/cookies.md).

</details>

hsc_gov_subscriber/utils/config.py
import os
import re
from configparser import ConfigParser
from datetime import datetime
from pathlib import Path

PROJECT_PATH = Path(__file__).parent.parent.resolve()

config_file = os.path.join(PROJECT_PATH, "config.ini")
cookies_file = os.path.join(PROJECT_PATH, "cookies.json")


class ConfigUnit:
    def __init__(self, section, name, value=None, default_value=None, description=None):
        self._section = section
        self._name = name
        self._value = value
        self._default_value = default_value
        self._description = description

    @property
    def section(self):
        return self._section

    @property
    def name(self):
        return self._name

    @property
    def default_value(self):
        return self._default_value

    @property
    def description(self):
        return self._description

    @property
    def value(self):
        return Config.get_raw(self.section, self.name)

    def set_value(self, val):
        self._value = val


class Config:
    _config = ConfigParser()
    _config.read(config_file)

    @classmethod
    def set(cls, section, key, value):
        cls._config.set(section, key, value)
        for var in vars(cls).values():
            if (
                    isinstance(var, ConfigUnit)
                    and var.section == section
                    and var.name == key
            ):
                var.set_value(value)

    @classmethod
    def get_raw(cls, section, key):
        return cls._config.get(section, key)

    @classmethod
    def get_units(cls):
        return [v for v in vars(cls).values() if isinstance(v, ConfigUnit)]

    @staticmethod
    def save():
        with open(config_file, "w") as f:
            Config._config.write(f)

    EMAIL = ConfigUnit(
        "application",
        "email",
        description="Пошта на яку прийде білет та нагадування.",
    )
    OFFICE_ID = ConfigUnit(
        "application",
        "office_id",
        description="ID офісу ГСЦ. (71 - ГСЦ м. Миколаїв, пров. Транспортний, 1а/1).",
    )
    QUESTION_ID = ConfigUnit(
        "application",
        "question_id",
        description="ID послуги. (56 - практичний іспит (транспортний засіб навчального закладу))",
    )
    SUB_ID = ConfigUnit(
        "application",
        "sub_id",
        description="ID запиту для практичний іспит (транспортний засіб навчального закладу)",
    )
    SCHOOL_VEHICLE_PRACTICAL_EXAM = ConfigUnit(
        "application",
        "school_vehicle_practical_exam",
        description="Транспортний засіб навчального закладу на практичний іспит",
    )
    VIN = ConfigUnit(
        "application",
        "vin",
        description="Номерний знак(AA1488BB) або VIN(останні 6 цифр)",
    )

    API_ID = ConfigUnit(
        "telethon",
        "api_id",
        description="App api_id з my.telegram.org"
    )
    API_HASH = ConfigUnit(
        "telethon",
        "api_hash",
        description="App api_hash з my.telegram.org"
    )

    START_DATE = ConfigUnit(
        "date",
        "start_date",
        description="дата з якої потрібно брати талон. (наприклад 2024-05-9 - якщо буде талон на 08.5.2024 - код не візьме такий талон, якщо прийде 09.5.2024 - код намагатиметься взяти талон)"
    )
    END_DATE = ConfigUnit(
        "date",
        "end_date",
        description="дата з якої вже не потрібно брати талон. (наприклад 2024-05-14 - якщо буде талон на 15.5.2024 - код не візьме такий талон, якщо прийде 14.5.2024 або 13.5.2024 - код намагатиметься взяти талон)"
    )


class ConfigException(Exception):
    pass


class ConfigValidation:
    VIN_PATTERN = r'([A-Za-z]{2}\d{4}[A-Za-z]{2})|(\d{6})'

    @staticmethod
    def _raise_config_exception(unit, message=None):
        raise ConfigException(
            f"Недійсне значення '{unit.value}' у розділі '{unit.section}.{unit.name}' у файлі конфігурації.\n" + message)

    @staticmethod
    def _validate_date_range(unit_start_date, unit_end_date):
        start_date = datetime.strptime(unit_start_date.value, "%Y-%m-%d")
        end_date = datetime.strptime(unit_end_date.value, "%Y-%m-%d")
        if not start_date <= end_date:
            ConfigValidation._raise_config_exception(unit_start_date,
                                                     f"{unit_start_date.section}.{unit_start_date.name} має бути меншим за {unit_end_date.section}.{unit_end_date.name}")

    @staticmethod
    def _check_not_none(unit: ConfigUnit):
        value = unit.value
        if (value is None
                or not value):
            ConfigValidation._raise_config_exception(unit, "Необхідно заповнити.")

    @classmethod
    def validate(cls):
        cls._validate_email()
        cls._validate_office_id()
        cls._validate_question_id()
        cls._validate_sub_id()
        cls._validate_school_vehicle_practical_exam()
        cls._validate_vin()

        cls._validate_api_id()
        cls._validate_api_hash()

        cls._validate_start_date()
        cls._validate_end_date()

    @staticmethod
    def _validate_email():
        ConfigValidation._check_not_none(Config.EMAIL)

    @staticmethod
    def _validate_office_id():
        ConfigValidation._check_not_none(Config.OFFICE_ID)

    @staticmethod
    def _validate_question_id():
        ConfigValidation._check_not_none(Config.QUESTION_ID)

    @staticmethod
    def _validate_sub_id():
        unit = Config.SUB_ID
        if int(Config.QUESTION_ID.value) == 56:
            ConfigValidation._check_not_none(unit)

    @staticmethod
    def _validate_school_vehicle_practical_exam():
        unit = Config.SCHOOL_VEHICLE_PRACTICAL_EXAM
        if int(Config.QUESTION_ID.value) == 56:
            ConfigValidation._check_not_none(unit)
            if bool(unit.value):
                ConfigValidation._check_not_none(unit)
                ConfigValidation._check_not_none(Config.SUB_ID)

    @staticmethod
    def _validate_vin():
        if int(Config.QUESTION_ID.value) == 49:
            unit = Config.VIN
            ConfigValidation._check_not_none(unit)
            if not re.search(ConfigValidation.VIN_PATTERN, unit.value):
                ConfigValidation._raise_config_exception(unit, unit.description)

    @staticmethod
    def _validate_api_id():
        ConfigValidation._check_not_none(Config.API_ID)

    @staticmethod
    def _validate_api_hash():
        ConfigValidation._check_not_none(Config.API_HASH)

    @staticmethod
    def _validate_start_date():
        ConfigValidation._check_not_none(Config.START_DATE)
        ConfigValidation._validate_date_range(Config.START_DATE, Config.END_DATE)

    @staticmethod
    def _validate_end_date():
        ConfigValidation._check_not_none(Config.END_DATE)
        ConfigValidation._validate_date_range(Config.START_DATE, Config.END_DATE)
